### PR TITLE
Support for Distelli

### PR DESCRIPTION
**Pull request by jorgenskogas**
Added support for Distelli build and deploy (ref #594)

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -7517,6 +7517,70 @@ function(data, match, end, ask) {
   }
 });
 
+// Distelli integration
+camp.route(/^\/distelli\/([^/]+)\/([^/]+)(?:\/(.+))?\.(svg|png|gif|jpg|json)$/, cache({
+  queryParams: ['token'],
+  handler: (data, match, sendBadge, request) => {
+    const type = match[1]; // build or deploy
```
Comment (paulmelnikow) [inline server.js]: Could you modify the regex so it accepts only these values?

File: server.js [hunk 2]
```diff
@@ -7517,6 +7517,70 @@ function(data, match, end, ask) {
...
+
+    request(apiUrl, {json: true}, function(err, res, data) {
+      try {
+        if (!res || err !== null || res.statusCode !== 200) {
+          badgeData.text[1] = 'inaccessible';
+          sendBadge(format, badgeData);
+          return;
+        }
+
+        const apiStatus = type === 'build' ? data.builds[0] : data.deployments[0].servers;
```
Comment (paulmelnikow) [inline server.js]: If you move this code inside the switch statement you wouldn't need the conditional. If you want to use the variable `apiStatus` in each one, you can declare it outside the switch, or else put braces around each case.

File: server.js [hunk 3]
```diff
@@ -7517,6 +7517,70 @@ function(data, match, end, ask) {
...
+              status = 'running';
+            } else if (apiStatus.done > 0) {
+              status = 'success';
+            }
+            break;
+          default:
+            break;
+        }
+
+        const message = status;
```
Comment (paulmelnikow) [inline server.js]: This line seems unnecessary. ^^

**Comment by paulmelnikow** [comment]
In addition to service tests, please add badge examples in `lib/all-badge-examples.js`.

File: server.js (hunk 1 above)
Comment (jorgenskogas) [reply server.js]: Sure 👍

File: server.js (hunk 3 above)
Comment (jorgenskogas) [reply server.js]: It's certainly is, not sure why I didn't remove it.

**Comment by jorgenskogas** [comment]
@paulmelnikow: I have done the changes you asked for. I wanted to do more on the service tests, but I have run out of time. Hope you will accept it so others in the community can do more on it later.

File: lib/all-badge-examples.js [hunk 4]
```diff
@@ -109,6 +109,16 @@ const allBadgeExamples = [
       name: 'Build'
     },
     examples: [
+      {
+        title: 'Distelli build',
+        previewUri: '/distelli/build/ORG/APPLICATION-NAME.svg?token=TOKEN',
+        exampleUri: '/distelli/build/ORG/APPLICATION-NAME.svg?token=TOKEN'
```
Comment (paulmelnikow) [inline lib/all-badge-examples.js]: The preview URI will render as a badge, so it would need a real org, app name, and token.

File: server.js [hunk 5]
```diff
@@ -7517,6 +7517,70 @@ function(data, match, end, ask) {
...
+        let status = 'unknown';
+
+        switch (type) {
+          case 'build':
+            status = data.builds[0].status.toLowerCase();
+            break;
+          case 'deploy':
+            var serverStatus = data.deployments[0].servers;
+            var hasErrors = serverStatus.failed > 0 ? true : false;
+            console.log(serverStatus);
```
Comment (paulmelnikow) [inline server.js]: Remove console.log please ^^

File: service-tests/distelli.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    value: Joi.equal(
+      'unknown',
+      'waiting',
+      'queued',
+      'running',
+      'success',
+      'failed',
+      'inaccessible'
+    )
+  }));
```
Comment (paulmelnikow) [inline service-tests/distelli.js]: For these tests to be useful, they should hit a working URL, and obtain a response that indicates the service code is working. A deploy that returns failed, running, success, and a build that returns whatever a real build should.

File: lib/all-badge-examples.js (hunk 4 above)
Comment (jorgenskogas) [reply lib/all-badge-examples.js]: I only have a dev subscription on Distelli, so I can not add more users. Not sure if the API support read only tokens, I will ask Distelli about this.

Comment (jorgenskogas) [reply lib/all-badge-examples.js]: Distelli only have tokens with full permissions.

Comment (paulmelnikow) [reply lib/all-badge-examples.js]: Hmm. Using full-access token is unfortunate. Of course anyone looking at the badge would see the token, plus sending it to the Shields server is a security risk.

Do you think a status-only token is something they might add?

**Comment by paulmelnikow** [comment]
Thanks for being the first to contribute [to our OpenCollective](https://opencollective.com/shields)! 😃 💵 💯

File: lib/all-badge-examples.js (hunk 4 above)
Comment (jorgenskogas) [reply lib/all-badge-examples.js]: I have asked them about this, but not received any feedback on this matter. I will update when I know more. They told me that status badges was on their roadmap.

**Comment by jorgenskogas** [comment]
My pleasure :)

File: lib/all-badge-examples.js (hunk 4 above)
Comment (jorgenskogas) [reply lib/all-badge-examples.js]: They have added RO tokens to their road map. ETA minimum 3 weeks, they will notify me.

Comment (paulmelnikow) [reply lib/all-badge-examples.js]: That is great news!

Comment (paulmelnikow) [reply lib/all-badge-examples.js]: How about I close this, and you can reopen when you hear from them?

**Comment by paulmelnikow** [comment]
Waiting for upstream help, see https://github.com/badges/shields/pull/1310#discussion_r154534015.

**Comment by ekeller** [comment]
@paulmelnikow I would like to finish this, they already have the token [token](https://puppet.com/docs/pipelines-for-apps/team/api-token.html#create-an-api-token) creation, problem is that it is not read-only, is this why this is being hold up?

Should we change name to Puppet now, since there's officially no more Distelli?

If we are to proceed we should be able to have at least two badges I can think of right now
- build status badge `Sucess/Failed`
- deployment badge `v76` i.e.

**Comment by mapoulin** [comment]
https://support.puppet.com/hc/en-us/articles/360024576474-KB-0347-Puppet-Pipelines-End-of-Life-notice

Puppet is discontinuing the product. RIP.

**Pull request closed without merging**